[core] rate limit handling on online file source

ivovandongen opened the pull request:
Fixes #5821

ivovandongen commented:
I've implemented the core and platform specifics. All platforms, expect for Android, have the same unit test. It would be nice to also test this against an actual server with a low rate limit as well, just to be sure.

- @jfirebaugh Can you take a look at the core stuff?
- @tmpsantos Same for linux/QT?
- @frederoni iOS/macOS?
- @zugaldia Android?

jfirebaugh commented on include/mbgl/storage/response.hpp:
```diff
@@ -45,15 +46,24 @@ class Response::Error {
+    Error(Reason, std::string = "", optional<RetryAfter> = {});
+    
+private:
+    optional<Timestamp> parseRetryAfter(const optional<RetryAfter> retryAfter_) const;
```
Can you move this to `http_header.hpp` and make the signature:

```
optional<Timestamp> parseRetryHeaders(const optional<std::string>& retryAfter,
                                      const optional<std::string>& xRateLimitReset);
```

Advantages:
* No need for `RetryAfter` type
* Factor conditional and integer conversion logic from individual SDK implementations

jfirebaugh commented on test/storage/online_file_source.cpp:
```diff
@@ -359,3 +359,56 @@ TEST(OnlineFileSource, TEST_REQUIRES_SERVER(NetworkStatusOnlineOffline)) {
     loop.run();
 }
+
+TEST(OnlineFileSource, TEST_REQUIRES_SERVER(RateLimit)) {
```
We shouldn't add any more timing-dependent tests to the test suite. The ones we have already spuriously fail intermittently (as will most any timing-dependent test, eventually).

Instead, how about making `errorRetryTimeout` public and unit testing it instead?

ivovandongen commented:
@jfirebaugh Thanks for taking a look. I've changed the implementations and will rework the tests.

ivovandongen commented:
@jfirebaugh 
- I've extracted the timeout functions to a header in src/util so it is testable, didn't want to add it to the public header as it is not quite public api. Let me know if this isn't what you had in mind.
- I kept the integration tests so that the platform bindings are tested as well.
- I've added unit tests on the timeout functions directly. 

When writing the unit tests I ran into a little issue where the the timeout became negative on high number of errors as the sign bit flipped ([here](https://github.com/mapbox/mapbox-gl-native/blob/e1a6795a73af844ff63d99b35fb4695177d9494d/platform/default/online_file_source.cpp#L206) among others). Changed all three exponential backoffs so that doesn't happen (`1` -> `1u`) [here](https://github.com/mapbox/mapbox-gl-native/blob/1eebf712623fb7f9cd85a6aac2ba124663216a84/src/mbgl/util/http_timeout.hpp#L14)

jfirebaugh commented on src/mbgl/util/http_header.cpp:
```diff
@@ -28,6 +28,24 @@ CacheControl CacheControl::parse(const std::string& value) {
+    
+optional<Timestamp> parseRetryHeaders(const optional<std::string>& retryAfter,
+                                      const optional<std::string>& xRateLimitReset) {
+    optional<Timestamp> result;
```
Prefer early returns to temporary variables:

```
if (!retryAfter) {
    return ...;
}

try {
    return ...;
} catch ... {
    return ...;
}
```

jfirebaugh commented on src/mbgl/util/http_header.cpp:
```diff
@@ -28,6 +28,24 @@ CacheControl CacheControl::parse(const std::string& value) {
+        try {
+            auto secs = std::chrono::seconds(std::stoi(*retryAfter));
+            result = std::chrono::time_point_cast<Seconds>(std::chrono::system_clock::now() + secs);
+        } catch (std::invalid_argument) {
```
`stoi` can also throw `std::out_of_range`. It's probably best to use `catch (...)` here.

jfirebaugh commented on src/mbgl/util/http_header.cpp:
```diff
@@ -28,6 +28,24 @@ CacheControl CacheControl::parse(const std::string& value) {
+            result = util::parseTimestamp((*retryAfter).c_str());
+        }
+    } else if(xRateLimitReset) {
+        result = util::parseTimestamp(std::stoi(*xRateLimitReset));
```
Need to handle exceptions from `std::stoi` here.

jfirebaugh replied:
Hmm, actually, do we want `parseRetryHeaders` to swallow malformed header values? I'm not sure. What does existing code do in the case of malformed `Cache-Control`, etc.?

jfirebaugh commented on src/mbgl/util/http_timeout.hpp:
```diff
@@ -0,0 +1,43 @@
+namespace mbgl {
+namespace http {
+
+static Duration errorRetryTimeout(Response::Error::Reason failedRequestReason, uint32_t failedRequests, optional<Timestamp> retryAfter = {}) {
```
Using an internal header for the method declaration is 👍 , but the definition should go in a sibling http_timeout.cpp file. And then remove `static` here.

jfirebaugh commented on src/mbgl/util/http_timeout.hpp:
```diff
@@ -0,0 +1,43 @@
+
+    if (failedRequestReason == Response::Error::Reason::Server) {
+        // Retry after one second three times, then start exponential backoff.
+        return Seconds(failedRequests <= 3 ? 1 : 1u << std::min(failedRequests - 3, 31u));
```
Nice catch on the signed integer overflow!

tmpsantos commented on src/mbgl/util/http_header.cpp:
```diff
@@ -28,6 +28,26 @@ CacheControl CacheControl::parse(const std::string& value) {
+    
+optional<Timestamp> parseRetryHeaders(const optional<std::string>& retryAfter,
+                                      const optional<std::string>& xRateLimitReset) {
+    if(retryAfter) {
```
nit, missing space after `if`

tmpsantos commented on src/mbgl/util/http_header.cpp:
```diff
@@ -28,6 +28,26 @@ CacheControl CacheControl::parse(const std::string& value) {
+        } catch (...) {
+            return util::parseTimestamp((*retryAfter).c_str());
+        }
+    } else if(xRateLimitReset) {
```
ditto.

tmpsantos commented on src/mbgl/util/http_timeout.cpp:
```diff
@@ -0,0 +1,39 @@
+            return *retryAfter - util::now();
+        } else {
+            //Default
+            return Seconds(5);
```
Move the magic constant to `./include/mbgl/util/constants.hpp`. We already have similar things like `CLOCK_SKEW_RETRY_TIMEOUT` there.

tmpsantos commented on platform/qt/src/http_request.cpp:
```diff
@@ -79,7 +82,12 @@ void HTTPRequest::handleNetworkReply(QNetworkReply *reply)
+        } else if (header == "x-rate-limit-reset") {
+            xRateLimitReset = std::string(line.second.constData(), line.second.size());
         }
+
```
Not needed.

tmpsantos commented on platform/default/http_file_source.cpp:
```diff
@@ -325,8 +329,13 @@ size_t HTTPRequest::headerCallback(char *const buffer, const size_t size, const
+        baton->xRateLimitReset = std::string(buffer + begin, length - begin - 2); // remove \r\n
     }
 
+
```
Not needed.

tmpsantos commented:
Thanks for adding more tests!

ivovandongen commented:
@tmpsantos Thanks for the review! Made the changes as suggested. Waiting on CI and then I'll merge.
